Cython/Utility/MemoryView_C.c

```diff
@@ -26,8 +26,59 @@ typedef struct {
 #endif
 
 #define __pyx_atomic_int_type int
+#define __pyx_nonatomic_int_type int
+
+// for standard C/C++ atomics, get the headers first so we have ATOMIC_INT_LOCK_FREE
+// defined when we decide to use them
+#if CYTHON_ATOMICS && (defined(__STDC_VERSION__) && \
+                        (__STDC_VERSION__ >= 201112L) && \
+                        !defined(__STDC_NO_ATOMICS__))
+    #include <stdatomic.h>
+#elif CYTHON_ATOMICS && (defined(__cplusplus) && ( \
+                    (__cplusplus >= 201103L) || \
+                    (defined(_MSC_VER) && _MSC_VER >= 1700)))
+    #include <atomic>
+#endif
 
-#if CYTHON_ATOMICS && (__GNUC__ >= 5 || (__GNUC__ == 4 && \
+#if CYTHON_ATOMICS && (defined(__STDC_VERSION__) && \
+                        (__STDC_VERSION__ >= 201112L) && \
+                        !defined(__STDC_NO_ATOMICS__) && \
+                       ATOMIC_INT_LOCK_FREE == 2)
+    /* C11 atomics are available
+       require ATOMIC_INT_LOCK_FREE because I'm nervous about the __pyx_atomic_int[2]
+       alignment trick in MemoryView.pyx if it uses mutexes
+     */
+    #undef __pyx_atomic_int_type
+    #define __pyx_atomic_int_type atomic_int
+    // TODO - it might be possible to use a less strict memory ordering here
+    #define __pyx_atomic_incr_aligned(value) atomic_fetch_add(value, 1)
+    #define __pyx_atomic_decr_aligned(value) atomic_fetch_sub(value, 1)
+    #if defined(__PYX_DEBUG_ATOMICS) && defined(_MSC_VER)
+        #pragma message ("Using standard C atomics")
+    #elif defined(__PYX_DEBUG_ATOMICS)
+        #warning "Using standard C atomics"
+    #endif
+#elif CYTHON_ATOMICS && (defined(__cplusplus) && ( \
+                    (__cplusplus >= 201103L) || \
+                    /*_MSC_VER 1700 is Visual Studio 2012 */ \
+                    (defined(_MSC_VER) && _MSC_VER >= 1700)) && \
+                    ATOMIC_INT_LOCK_FREE == 2)
+    /* C++11 atomics are available
+       require ATOMIC_INT_LOCK_FREE because I'm nervous about the __pyx_atomic_int[2]
+       alignment trick in MemoryView.pyx if it uses mutexes
+     */
```

Comment:
```suggestion
    // C++11 atomics are available.
    // Require ATOMIC_INT_LOCK_FREE because I'm nervous about the __pyx_atomic_int[2]
    // alignment trick in MemoryView.pyx if it uses mutexes.
```